### PR TITLE
PM5 update

### PR DESCRIPTION
**Pull request by NopeNotDark**

### CODE REVIEW

**Comment by NopeNotDark** [comment]
Needs testing.

File: src/muqsit/arenagenerator/ArenaAllocator.php [hunk 1]
```diff
@@ -4,55 +4,50 @@
...
-	}
-
-	public function reallocate() : void{
-		$this->next_x = 0;
-		$this->next_z = 0;
-		++$this->allocations;
-	}
+final class ArenaAllocator
+{
+    public static function optimized(): self {
```
Comment (Muqsit) [inline src/muqsit/arenagenerator/ArenaAllocator.php:16]: The code needs to match the original format (indentation, brackets, etc.).

File: src/muqsit/arenagenerator/layout/chunk/PreloadedChunk.php [hunk 2]
```diff
@@ -4,90 +4,66 @@
...
 	 * @param SubChunk[] $sub_chunks
-	 * @param CompoundTag[] $entities
-	 * @param CompoundTag[] $tiles
-	 * @param string $biome_ids
+	 * @param Tile[] $tiles
 	 * @param int[] $height_map
 	 * @param bool|null $light_populated
 	 */
-	public function __construct(array $sub_chunks, array $entities, array $tiles, string $biome_ids, array $height_map, ?bool $light_populated){
+	public function __construct(array $sub_chunks, array $tiles, array $height_map, ?bool $light_populated){
```
Comment (Muqsit) [inline src/muqsit/arenagenerator/layout/chunk/PreloadedChunk.php:42]: Existing entities must be transferred onto copied arenas (this change will prevent stuff like zombies in a region or hung paintings to be lost during copying).

File: src/muqsit/arenagenerator/layout/chunk/PreloadedChunk.php [hunk 3]
```diff
@@ -4,90 +4,66 @@
...
-			// $world->addEntity($entity_factory->createFromData($world, $data)); TODO: Find out why this sometimes results in "Entity 645 has already been added to this world"
-		}
-
 		$tile_factory = TileFactory::getInstance();
 		foreach($this->tiles as $data){
 			$data = clone $data;
-			$movers->block_entity->move($chunk_x, $chunk_z, $data);
-			$tile = $tile_factory->createFromData($world, $data);
+			$movers->block_entity->move($chunk_x, $chunk_z, $data->getCleanedNBT());
+			$tile = $tile_factory->createFromData($world, $data->getCleanedNBT());
```
Comment (Muqsit) [inline src/muqsit/arenagenerator/layout/chunk/PreloadedChunk.php:66]: `Tile::getSaveNbt()` would give the complete NBT to restore tiles to their exact state.

**Review by Muqsit — CHANGES_REQUESTED** [review]

**Pull request closed without merging**